src/main/java/org/broadinstitute/hellbender/tools/spark/sv/discovery/SvDiscoverFromLocalAssemblyContigAlignmentsSpark.java

```diff
@@ -147,52 +148,46 @@ protected void runTool(final JavaSparkContext ctx) {
-            this.misassembly = misassembly;
-            this.simple = simple;
-            this.complex = complex;
+        private AssemblyContigsClassifiedByAlignmentSignatures(final JavaRDD<AssemblyContigWithFineTunedAlignments> contigs) {
+            unknown = contigs.filter(tig -> tig.getAlignmentSignatureBasicType().equals(UNKNOWN)).cache();
+            simple = contigs.filter(tig -> tig.getAlignmentSignatureBasicType().equals(SIMPLE_CHIMERA)).cache();
+            complex = contigs.filter(tig -> tig.getAlignmentSignatureBasicType().equals(COMPLEX)).cache();
         }
 
         /**
          * Write SAM file, if requested, for original alignments of contigs recognized as "Ambiguous", "Incomplete", and "MisAssemblySuspect"
          * TODO: 11/17/17 salvation on assembly contigs that 1) has ambiguous "best" configuration, and 2) has incomplete picture; and flag accordingly
-         *
-         * TODO: 3/4/18 a bug is present here that even though only one alignment has not-bad MQ, it could contain a large gap,
-         *      depending on the behavior of the other gap-less bad mappings,
-         *      we may end up classifying the whole contig as incomplete, or signal-less,
-         *      we should keep the single not-bad mapping and annotate accordingly
          */
-        private void writeSAMfilesForSuspicious(final String outputPrefix, final JavaRDD<GATKRead> assemblyRawAlignments,
-                                                final SAMFileHeader header) {
-            final Set<String> ambiguousContigNames = new HashSet<>( ambiguous.map(AssemblyContigWithFineTunedAlignments::getContigName).collect() );
-            final Set<String> incompleteContigNames = new HashSet<>( incomplete.map(AssemblyContigWithFineTunedAlignments::getContigName).collect() );
-            final Set<String> misAssemblySuspectContigNames = new HashSet<>( misassembly.map(AssemblyContigWithFineTunedAlignments::getContigName).collect() );
-
-            SvDiscoveryUtils.writeSAMRecords(assemblyRawAlignments, ambiguousContigNames, outputPrefix + AssemblyContigWithFineTunedAlignments.ReasonForAlignmentClassificationFailure.AMBIGUOUS.name() + ".bam", header);
-            SvDiscoveryUtils.writeSAMRecords(assemblyRawAlignments, incompleteContigNames, outputPrefix + AssemblyContigWithFineTunedAlignments.ReasonForAlignmentClassificationFailure.INCOMPLETE.name() + ".bam", header);
-            SvDiscoveryUtils.writeSAMRecords(assemblyRawAlignments, misAssemblySuspectContigNames, outputPrefix + AssemblyContigWithFineTunedAlignments.ReasonForAlignmentClassificationFailure.MIS_ASSEMBLY_OR_MAPPING_SUSPECT.name() + ".bam", header);
+        private void writeSAMfilesForUnknown(final String outputPrefix, final JavaRDD<GATKRead> assemblyRawAlignments,
+                                             final SAMFileHeader header) {
+
+            final Map<ReasonForAlignmentClassificationFailure, Iterable<String>> reasonToContigNames =
```

Comment:
This is a better but you are still collecting the RDD and passing over the collection three times. What I meant by my original suggestion was this: Why not make the map go the other way, ie make a `Map<String, ReasonForAlignmentClassificationFailure>` that maps contig names to their reasons? Then make a `Map<ReasonForAlignmentClassificationFailure, SAMFileWriter>` with three entries. Then you only have to iterate over the collection of reads once to write everything out (you just look up the writer for each entry).